pkg/gadgets/internal/test/runner.go

```diff
@@ -0,0 +1,183 @@
+	Comm      string
+	UID       int
+	MountNsID uint64
+}
+
+// Runner is a helper type to execute tests in different conditions. It
+// creates a go routine that is executed in a different mount namespace,
+// user ID, etc. to simulate events hapenning inside containers.
+type Runner struct {
+	config *RunnerConfig
+
+	tasks   chan func() error
+	replies chan error
+	Info    *RunnerInfo
+}
+
+func NewRunner(config *RunnerConfig) (*Runner, error) {
+	if config == nil {
+		config = &RunnerConfig{}
+	}
+	r := &Runner{
+		config:  config,
+		tasks:   make(chan func() error),
+		replies: make(chan error),
+	}
+
+	go r.runLoop()
+
+	if err := <-r.replies; err != nil {
+		r.Close()
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func (r *Runner) Run(f func() error) error {
+	r.tasks <- f
+	return <-r.replies
+}
+
+func (r *Runner) Close() {
+	if r.tasks != nil {
+		close(r.tasks)
+		r.tasks = nil
+	}
+	if r.replies != nil {
+		close(r.replies)
+		r.replies = nil
+	}
+}
+
+func (r *Runner) runLoop() {
+	// Don't switch thread in the following code
+	runtime.LockOSThread()
+	// We aren't calling runtime.UnlockOSThread() to let this thread
+	// die because some changes are not recoverable, like switching
+	// to a non-root UID.
+
+	mntns, err := createMntNamespace()
+	if err != nil {
+		r.replies <- fmt.Errorf("creating mount namespace: %w", err)
+		return
+	}
+	defer unix.Close(mntns)
+
+	if r.config.UID != 0 {
+		// syscall.Setuid() can't be used here because it'll
+		// change the UID of all threads and we only need to
+		// change the one of this thread.
+		// https://github.com/golang/go/commit/d1b1145cace8b968307f9311ff611e4bb810710c
+		_, _, errno := syscall.Syscall(syscall.SYS_SETUID, uintptr(r.config.UID), 0, 0)
+		if errno != 0 {
+			r.replies <- fmt.Errorf("setting uid: %w", err)
+			return
+		}
+	}
+
+	pid := os.Getpid()
+	tid := unix.Gettid()
+
+	mountnsid, err := getMntNamespaceInode(pid, tid)
+	if err != nil {
+		r.replies <- fmt.Errorf("getting namespace inode: %w", err)
+		return
+	}
+
+	comm, err := os.Executable()
+	if err != nil {
+		r.replies <- fmt.Errorf("getting current executable: %w", err)
+		return
+	}
+
+	r.Info = &RunnerInfo{
+		Pid:       pid,
+		Comm:      filepath.Base(comm),
+		UID:       r.config.UID,
+		MountNsID: mountnsid,
+	}
+
+	// Indicate it's ready to process tasks
+	r.replies <- nil
+
+	for task := range r.tasks {
+		r.replies <- task()
+	}
+}
+
+func createMntNamespace() (int, error) {
+	if err := unix.Unshare(syscall.CLONE_NEWNS); err != nil {
+		return -1, err
+	}
+	return getMntNsHandle()
+}
+
+func getMntNamespaceInode(pid, tid int) (uint64, error) {
+	fileinfo, err := os.Stat(fmt.Sprintf("/proc/%d/task/%d/ns/mnt", pid, tid))
+	if err != nil {
+		return 0, err
+	}
+	stat, ok := fileinfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return 0, errors.New("not a syscall.Stat_t")
+	}
+	return stat.Ino, nil
+}
+
+// getMntNsHandle gets a handle to the current thread mount namespace.
+func getMntNsHandle() (int, error) {
```

Comment:
What is the point of this function? It seems the handle is never used.

Reply:
There is not any reason. Probably it's a leftover after so many rework. I opened https://github.com/kinvolk/inspektor-gadget/pull/992 to remove it.